templates/fine-tune-llama2/train.py

```diff
@@ -0,0 +1,871 @@
+    ray.init(
+        runtime_env={
+            "env_vars": {
+                "HF_HOME": "/mnt/local_storage/.cache/huggingface",
+                "RAY_AIR_LOCAL_CACHE_DIR": os.environ["RAY_AIR_LOCAL_CACHE_DIR"],
+            },
+            # This is not needed to run in Workspace. However, it might be needed for other environments.
+            # "working_dir": ".",
+        }
+    )
+
+    # Read data
+    train_ds = ray.data.read_json(config["train_path"])
+    if args.test_path is not None:
+        valid_ds = ray.data.read_json(config["test_path"])
+    else:
+        valid_ds = None
+
+    # Read special tokens json file
+    with open(config["special_token_path"], "r") as json_file:
+        special_tokens = json.load(json_file)["tokens"]
+
+    # Specify the cloud bucket path to use when saving training artifacts
+    # ANYSCALE_ARTIFACT_STORAGE is the default cloud storage bucket created for Anyscale users
+    artifact_storage = os.environ.get("ANYSCALE_ARTIFACT_STORAGE", "artifact_storage")
+    user_name = re.sub(r"\s+", "__", os.environ.get("ANYSCALE_USERNAME", "user"))
+    model_name=config.get("model_name")
+    storage_path = (
+        f"{artifact_storage}/{user_name}/ft_llms_with_deepspeed/{model_name}"
+    )
+
+    # trial_name = f"{model_name}".split("/")[-1]
+    # if config["lora"]:
+    #     trial_name += "-lora"
+
+    # set up Ray Train trainer
+    trainer = TorchTrainer(
+        training_function,
+        train_loop_config={
+            "config": config,
+            "special_tokens": special_tokens,
+        },
+        run_config=train.RunConfig(
+            storage_path=storage_path,
+            checkpoint_config=train.CheckpointConfig(
+                num_to_keep=config["num_checkpoints_to_keep"],
+                checkpoint_score_attribute="perplexity",
+                checkpoint_score_order="min",
+            ),
+        ),
+        scaling_config=train.ScalingConfig(
+            num_workers=config["num_devices"],
+            use_gpu=True,
+            resources_per_worker={"GPU": 1},
+        ),
+        datasets={"train": train_ds, "valid": valid_ds},
+        dataset_config=ray.train.DataConfig(datasets_to_split=["train", "valid"]),
+    )
+
+    try:
+        # Start training
+        result: train.Result = trainer.fit()
+        # `best_checkpoints` are sorted in increasing score order.
+        # (Ex: in this case, negative perplexity, since we set `checkpoint_score_order=min`)
+        best_checkpoint, best_checkpoint_metrics = result.best_checkpoints[-1]
+
+        print("Results are stored at:")
+        print(result.path)
+        print("Best checkpoint is stored at:")
+        print(best_checkpoint)
+        print(f"With perplexity: {best_checkpoint_metrics['perplexity']}")
+
+        print("Process completed.")
+
+    except Exception as e:
+        print("Failed to fine-tune the model. Exiting...")
+        print(f"Error: {e}")
+        exit()  # This will stop the script
```

Comment:
We can also use `sys.exit()` here. At least for consistency but it's also the intended pattern.

Reply:
Done.